File: internal/domain/workflow.go
package domain

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/samber/lo"

	"github.com/certimate-go/certimate/internal/domain/expr"
	xmaps "github.com/certimate-go/certimate/pkg/utils/maps"
)

const CollectionNameWorkflow = "workflow"

type Workflow struct {
	Meta
	Name          string                `db:"name"          json:"name"`
	Description   string                `db:"description"   json:"description"`
	Trigger       WorkflowTriggerType   `db:"trigger"       json:"trigger"`
	TriggerCron   string                `db:"triggerCron"   json:"triggerCron"`
	Enabled       bool                  `db:"enabled"       json:"enabled"`
	GraphDraft    *WorkflowGraph        `db:"graphDraft"    json:"graphDraft"`
	GraphContent  *WorkflowGraph        `db:"graphContent"  json:"graphContent"`
	HasDraft      bool                  `db:"hasDraft"      json:"hasDraft"`
	HasContent    bool                  `db:"hasContent"    json:"hasContent"`
	LastRunId     string                `db:"lastRunRef"    json:"lastRunId"`
	LastRunStatus WorkflowRunStatusType `db:"lastRunStatus" json:"lastRunStatus"`
	LastRunTime   time.Time             `db:"lastRunTime"   json:"lastRunTime"`
}

type WorkflowGraph struct {
	Nodes []*WorkflowNode `json:"nodes"`
}

func (g *WorkflowGraph) GetNodeById(nodeId string) (*WorkflowNode, bool) {
	return g.getNodeInBlocksById(g.Nodes, nodeId)
}

func (g *WorkflowGraph) getNodeInBlocksById(blocks []*WorkflowNode, nodeId string) (*WorkflowNode, bool) {
	for _, node := range blocks {
		if node.Id == nodeId {
			return node, true
		}

		if len(node.Blocks) > 0 {
			if found, ok := g.getNodeInBlocksById(node.Blocks, nodeId); ok {
				return found, true
			}
		}
	}

	return nil, false
}

func (g *WorkflowGraph) Verify() error {
	if len(g.Nodes) < 2 {
		return fmt.Errorf("invalid nodes length of graph")
	} else if g.Nodes[0].Type != WorkflowNodeTypeStart {
		return fmt.Errorf("the first node is not a start node")
	} else if g.Nodes[len(g.Nodes)-1].Type != WorkflowNodeTypeEnd {
		return fmt.Errorf("the last node is not an end node")
	}

	return nil
}

func (g *WorkflowGraph) Clone() *WorkflowGraph {
	return &WorkflowGraph{
		Nodes: g.Nodes,
	}
}

type WorkflowTriggerType string

const (
	WorkflowTriggerTypeScheduled = WorkflowTriggerType("scheduled")
	WorkflowTriggerTypeManual    = WorkflowTriggerType("manual")
)

type WorkflowNode struct {
	Id     string           `json:"id"` // 节点 ID 只在该工作流中唯一，在全局中不保证唯一性
	Type   WorkflowNodeType `json:"type"`
	Data   WorkflowNodeData `json:"data"`
	Blocks []*WorkflowNode  `json:"blocks,omitempty"`
}

type WorkflowNodeType string

const (
	WorkflowNodeTypeStart       = WorkflowNodeType("start")
	WorkflowNodeTypeEnd         = WorkflowNodeType("end")
	WorkflowNodeTypeCondition   = WorkflowNodeType("condition")
	WorkflowNodeTypeBranchBlock = WorkflowNodeType("branchBlock")
	WorkflowNodeTypeTryCatch    = WorkflowNodeType("tryCatch")
	WorkflowNodeTypeTryBlock    = WorkflowNodeType("tryBlock")
	WorkflowNodeTypeCatchBlock  = WorkflowNodeType("catchBlock")
	WorkflowNodeTypeDelay       = WorkflowNodeType("delay")
	WorkflowNodeTypeBizApply    = WorkflowNodeType("bizApply")
	WorkflowNodeTypeBizUpload   = WorkflowNodeType("bizUpload")
	WorkflowNodeTypeBizMonitor  = WorkflowNodeType("bizMonitor")
	WorkflowNodeTypeBizDeploy   = WorkflowNodeType("bizDeploy")
	WorkflowNodeTypeBizNotify   = WorkflowNodeType("bizNotify")
)

type WorkflowNodeData struct {
	Name     string             `json:"name"`
	Disabled bool               `json:"disabled,omitempty,omitzero"`
	Config   WorkflowNodeConfig `json:"config,omitempty,omitzero"`
}

type WorkflowNodeConfig map[string]any

func (c WorkflowNodeConfig) AsDelay() WorkflowNodeConfigForDelay {
	return WorkflowNodeConfigForDelay{
		Wait: xmaps.GetInt(c, "wait"),
	}
}

func (c WorkflowNodeConfig) AsBranchBlock() WorkflowNodeConfigForBranchBlock {
	expression := c["expression"]
	if expression == nil {
		return WorkflowNodeConfigForBranchBlock{}
	}

	exprRaw, _ := json.Marshal(expression)
	expr, err := expr.UnmarshalExpr([]byte(exprRaw))
	if err != nil {
		return WorkflowNodeConfigForBranchBlock{}
	}

	return WorkflowNodeConfigForBranchBlock{
		Expression: expr,
	}
}

func (c WorkflowNodeConfig) AsBizApply() WorkflowNodeConfigForBizApply {
	domains := lo.Filter(strings.Split(xmaps.GetString(c, "domains"), ";"), func(s string, _ int) bool { return s != "" })
	ipaddrs := lo.Filter(strings.Split(xmaps.GetString(c, "ipaddrs"), ";"), func(s string, _ int) bool { return s != "" })
	nameservers := lo.Filter(strings.Split(xmaps.GetString(c, "nameservers"), ";"), func(s string, _ int) bool { return s != "" })

	return WorkflowNodeConfigForBizApply{
		Domains:               domains,
		IPAddrs:               ipaddrs,
		ContactEmail:          xmaps.GetString(c, "contactEmail"),
		ChallengeType:         xmaps.GetString(c, "challengeType"),
		Provider:              xmaps.GetString(c, "provider"),
		ProviderAccessId:      xmaps.GetString(c, "providerAccessId"),
		ProviderConfig:        xmaps.GetKVMapAny(c, "providerConfig"),
		KeySource:             xmaps.GetOrDefaultString(c, "keySource", "auto"),
		KeyAlgorithm:          xmaps.GetOrDefaultString(c, "keyAlgorithm", string(CertificateKeyAlgorithmTypeRSA2048)),
		KeyContent:            xmaps.GetString(c, "keyContent"),
		CAProvider:            xmaps.GetString(c, "caProvider"),
		CAProviderAccessId:    xmaps.GetString(c, "caProviderAccessId"),
		CAProviderConfig:      xmaps.GetKVMapAny(c, "caProviderConfig"),
		ValidityLifetime:      xmaps.GetString(c, "validityLifetime"),
		PreferredChain:        xmaps.GetString(c, "preferredChain"),
		ACMEProfile:           xmaps.GetString(c, "acmeProfile"),
		Nameservers:           nameservers,
		DnsPropagationWait:    xmaps.GetInt(c, "dnsPropagationWait"),
		DnsPropagationTimeout: xmaps.GetInt(c, "dnsPropagationTimeout"),
		DnsTTL:                xmaps.GetInt(c, "dnsTTL"),
		HttpDelayWait:         xmaps.GetInt(c, "httpDelayWait"),
		DisableCommonName:     xmaps.GetBool(c, "disableCommonName"),
		DisableFollowCNAME:    xmaps.GetBool(c, "disableFollowCNAME"),
		DisableARI:            xmaps.GetBool(c, "disableARI"),
		SkipBeforeExpiryDays:  xmaps.GetInt(c, "skipBeforeExpiryDays"),
	}
}

func (c WorkflowNodeConfig) AsBizUpload() WorkflowNodeConfigForBizUpload {
	return WorkflowNodeConfigForBizUpload{
		Source:      xmaps.GetOrDefaultString(c, "source", "form"),
		Certificate: xmaps.GetString(c, "certificate"),
		PrivateKey:  xmaps.GetString(c, "privateKey"),
	}
}

func (c WorkflowNodeConfig) AsBizMonitor() WorkflowNodeConfigForBizMonitor {
	host := xmaps.GetString(c, "host")
	return WorkflowNodeConfigForBizMonitor{
		Host:        host,
		Port:        xmaps.GetOrDefaultInt32(c, "port", 443),
		Domain:      xmaps.GetOrDefaultString(c, "domain", host),
		RequestPath: xmaps.GetString(c, "path"),
	}
}

func (c WorkflowNodeConfig) AsBizDeploy() WorkflowNodeConfigForBizDeploy {
	return WorkflowNodeConfigForBizDeploy{
		CertificateOutputNodeId: xmaps.GetString(c, "certificateOutputNodeId"),
		Provider:                xmaps.GetString(c, "provider"),
		ProviderAccessId:        xmaps.GetString(c, "providerAccessId"),
		ProviderConfig:          xmaps.GetKVMapAny(c, "providerConfig"),
		SkipOnLastSucceeded:     xmaps.GetBool(c, "skipOnLastSucceeded"),
	}
}

func (c WorkflowNodeConfig) AsBizNotify() WorkflowNodeConfigForBizNotify {
	return WorkflowNodeConfigForBizNotify{
		Provider:             xmaps.GetString(c, "provider"),
		ProviderAccessId:     xmaps.GetString(c, "providerAccessId"),
		ProviderConfig:       xmaps.GetKVMapAny(c, "providerConfig"),
		Subject:              xmaps.GetString(c, "subject"),
		Message:              xmaps.GetString(c, "message"),
		SkipOnAllPrevSkipped: xmaps.GetBool(c, "skipOnAllPrevSkipped"),
	}
}

type WorkflowNodeConfigForDelay struct {
	Wait int `json:"wait"` // 等待时间
}

type WorkflowNodeConfigForBranchBlock struct {
	Expression expr.Expr `json:"expression"` // 条件表达式
}

type WorkflowNodeConfigForBizApply struct {
	Domains               []string       `json:"domains"`                         // 域名列表，以半角分号分隔
	IPAddrs               []string       `json:"ipaddrs"`                         // IP 地址列表，以半角分号分隔
	ContactEmail          string         `json:"contactEmail"`                    // 联系邮箱
	ChallengeType         string         `json:"challengeType"`                   // 质询方式
	Provider              string         `json:"provider"`                        // 质询提供商
	ProviderAccessId      string         `json:"providerAccessId"`                // 质询提供商授权记录 ID
	ProviderConfig        map[string]any `json:"providerConfig,omitempty"`        // 质询提供商额外配置
	CAProvider            string         `json:"caProvider,omitempty"`            // CA 提供商（零值时使用全局配置）
	CAProviderAccessId    string         `json:"caProviderAccessId,omitempty"`    // CA 提供商授权记录 ID
	CAProviderConfig      map[string]any `json:"caProviderConfig,omitempty"`      // CA 提供商额外配置
	KeySource             string         `json:"keySource"`                       // 私钥来源，可取值 "auto"、"reuse"、"custom"（零值时默认值 "auto"）
	KeyAlgorithm          string         `json:"keyAlgorithm,omitempty"`          // 私钥算法
	KeyContent            string         `json:"keyContent,omitempty"`            // 私钥内容
	ValidityLifetime      string         `json:"validityLifetime,omitempty"`      // 有效期，形如 "30d"、"6h"
	PreferredChain        string         `json:"preferredChain,omitempty"`        // 首选证书链
	ACMEProfile           string         `json:"acmeProfile,omitempty"`           // ACME Profiles Extension
	Nameservers           []string       `json:"nameservers,omitempty"`           // DNS 服务器列表，以半角分号分隔。等同于 lego 的 `--dns.resolvers` 参数
	DnsPropagationWait    int            `json:"dnsPropagationWait,omitempty"`    // DNS 传播等待时间。等同于 lego 的 `--dns.propagation-wait` 参数
	DnsPropagationTimeout int            `json:"dnsPropagationTimeout,omitempty"` // DNS 传播检查超时时间。等同于 lego 的 `--dns-timeout` 参数
	DnsTTL                int            `json:"dnsTTL,omitempty"`                // DNS 解析记录 TTL
	HttpDelayWait         int            `json:"httpDelayWait,omitempty"`         // HTTP 等待时间。等同于 lego 的 `--http.delay` 参数
	DisableCommonName     bool           `json:"disableCommonName,omitempty"`     // 是否不包含 CommonName。等同于 lego 的 `--disable-cn` 参数
	DisableFollowCNAME    bool           `json:"disableFollowCNAME,omitempty"`    // 是否关闭 CNAME 跟随
	DisableARI            bool           `json:"disableARI,omitempty"`            // 是否关闭 ARI
	SkipBeforeExpiryDays  int            `json:"skipBeforeExpiryDays,omitempty"`  // 证书到期前多少天前跳过续期
}

type WorkflowNodeConfigForBizUpload struct {
	Source      string `json:"source"`      // 证书来源，可取值 "form"、"local"、"url"（零值时默认值 "form"）
	Certificate string `json:"certificate"` // 证书，根据证书来源决定是 PEM 内容 / 文件路径 / URL
	PrivateKey  string `json:"privateKey"`  // 私钥，根据证书来源决定是 PEM 内容 / 文件路径 / URL
}

type WorkflowNodeConfigForBizMonitor struct {
	Host        string `json:"host"`                  // 主机地址
	Port        int32  `json:"port,omitempty"`        // 端口（零值时默认值 443）
	Domain      string `json:"domain,omitempty"`      // 域名（零值时默认值 [Host]）
	RequestPath string `json:"requestPath,omitempty"` // 请求路径
}

type WorkflowNodeConfigForBizDeploy struct {
	CertificateOutputNodeId string         `json:"certificateOutputNodeId"`    // 前序证书输出节点 ID
	Provider                string         `json:"provider"`                   // 主机提供商
	ProviderAccessId        string         `json:"providerAccessId,omitempty"` // 主机提供商授权记录 ID
	ProviderConfig          map[string]any `json:"providerConfig,omitempty"`   // 主机提供商额外配置
	SkipOnLastSucceeded     bool           `json:"skipOnLastSucceeded"`        // 上次部署成功时是否跳过
}

type WorkflowNodeConfigForBizNotify struct {
	Provider             string         `json:"provider"`                 // 通知提供商
	ProviderAccessId     string         `json:"providerAccessId"`         // 通知提供商授权记录 ID
	ProviderConfig       map[string]any `json:"providerConfig,omitempty"` // 通知提供商额外配置
	Subject              string         `json:"subject"`                  // 通知主题
	Message              string         `json:"message"`                  // 通知内容
	SkipOnAllPrevSkipped bool           `json:"skipOnAllPrevSkipped"`     // 前序节点均已跳过时是否跳过
}

File: ui/src/domain/workflow.ts
import { getI18n } from "react-i18next";
import { Immer } from "immer";
import { nanoid } from "nanoid";

import { type WorkflowRunModel } from "./workflowRun";

export interface WorkflowModel extends BaseModel {
  name: string;
  description?: string;
  trigger: string;
  triggerCron?: string;
  enabled?: boolean;
  graphDraft?: WorkflowGraph;
  graphContent?: WorkflowGraph;
  hasDraft?: boolean;
  hasContent?: boolean;
  lastRunRef?: string;
  lastRunStatus?: string;
  lastRunTime?: string;
  expand?: {
    lastRunRef?: Pick<WorkflowRunModel, "id" | "status" | "trigger" | "startedAt" | "endedAt" | "error">;
  };
}

export interface WorkflowGraph {
  nodes: WorkflowNode[];
}

export const WORKFLOW_TRIGGERS = Object.freeze({
  SCHEDULED: "scheduled",
  MANUAL: "manual",
} as const);

export type WorkflowTriggerType = (typeof WORKFLOW_TRIGGERS)[keyof typeof WORKFLOW_TRIGGERS];

// #region Node
export const WORKFLOW_NODE_TYPES = Object.freeze({
  START: "start",
  END: "end",
  DELAY: "delay",
  CONDITION: "condition",
  BRANCHBLOCK: "branchBlock",
  TRYCATCH: "tryCatch",
  TRYBLOCK: "tryBlock",
  CATCHBLOCK: "catchBlock",
  BIZ_APPLY: "bizApply",
  BIZ_UPLOAD: "bizUpload",
  BIZ_MONITOR: "bizMonitor",
  BIZ_DEPLOY: "bizDeploy",
  BIZ_NOTIFY: "bizNotify",
} as const);

export type WorkflowNodeType = (typeof WORKFLOW_NODE_TYPES)[keyof typeof WORKFLOW_NODE_TYPES];

export type WorkflowNode = {
  id: string;
  type: WorkflowNodeType;
  data: {
    name?: string;
    disabled?: boolean;
    config?: Record<string, unknown>;
    [key: string]: unknown;
  };
  blocks?: WorkflowNode[];
};

export type WorkflowNodeConfigForStart = {
  trigger: string;
  triggerCron?: string;
};

export const defaultNodeConfigForStart = (): Partial<WorkflowNodeConfigForStart> => {
  return {
    trigger: WORKFLOW_TRIGGERS.MANUAL,
  };
};

export type WorkflowNodeConfigForDelay = {
  wait?: number;
};

export const defaultNodeConfigForDelay = (): Partial<WorkflowNodeConfigForDelay> => {
  return {};
};

export type WorkflowNodeConfigForBranchBlock = {
  expression?: Expr;
};

export const defaultNodeConfigForBranchBlock = (): Partial<WorkflowNodeConfigForBranchBlock> => {
  return {};
};

export type WorkflowNodeConfigForBizApply = {
  identifier: "domain" | "ip";
  domains: string;
  ipaddrs: string;
  contactEmail: string;
  challengeType: string;
  provider: string;
  providerAccessId: string;
  providerConfig?: Record<string, unknown>;
  caProvider?: string;
  caProviderAccessId?: string;
  caProviderConfig?: Record<string, unknown>;
  keySource: string;
  keyAlgorithm: string;
  keyContent?: string;
  validityLifetime?: string;
  acmeProfile?: string;
  nameservers?: string;
  dnsPropagationWait?: number;
  dnsPropagationTimeout?: number;
  dnsTTL?: number;
  httpDelayWait?: number;
  disableFollowCNAME?: boolean;
  disableARI?: boolean;
  skipBeforeExpiryDays: number;
};

export const defaultNodeConfigForBizApply = (): Partial<WorkflowNodeConfigForBizApply> => {
  return {
    challengeType: "dns-01" as const,
    keySource: "auto" as const,
    keyAlgorithm: "RSA2048" as const,
    skipBeforeExpiryDays: 30,
  };
};

export type WorkflowNodeConfigForBizUpload = {
  source: string;
  certificate: string;
  privateKey: string;
};

export const defaultNodeConfigForBizUpload = (): Partial<WorkflowNodeConfigForBizUpload> => {
  return {
    source: "form" as const,
  };
};

export type WorkflowNodeConfigForBizMonitor = {
  host: string;
  port: number;
  domain?: string;
  requestPath?: string;
};

export const defaultNodeConfigForBizMonitor = (): Partial<WorkflowNodeConfigForBizMonitor> => {
  return {
    host: "",
    port: 443,
    requestPath: "/",
  };
};

export type WorkflowNodeConfigForBizDeploy = {
  certificateOutputNodeId: string;
  provider: string;
  providerAccessId?: string;
  providerConfig?: Record<string, unknown>;
  skipOnLastSucceeded: boolean;
};

export const defaultNodeConfigForBizDeploy = (): Partial<WorkflowNodeConfigForBizDeploy> => {
  return {
    skipOnLastSucceeded: true,
  };
};

export type WorkflowNodeConfigForBizNotify = {
  subject: string;
  message: string;
  provider: string;
  providerAccessId: string;
  providerConfig?: Record<string, unknown>;
  skipOnAllPrevSkipped?: boolean;
};

export const defaultNodeConfigForBizNotify = (): Partial<WorkflowNodeConfigForBizNotify> => {
  return {};
};

export const newNodeId = (): string => {
  return nanoid()
    .replace(/^[_-]+/g, "")
    .replace(/[_-]+$/g, "");
};

export const newNode = (type: WorkflowNodeType, { i18n = getI18n() }: { i18n?: ReturnType<typeof getI18n> }): WorkflowNode => {
  const { t } = i18n;

  switch (type) {
    case WORKFLOW_NODE_TYPES.START:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.start.default_name"),
          config: defaultNodeConfigForStart(),
        },
      };

    case WORKFLOW_NODE_TYPES.END:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.end.default_name"),
        },
      };

    case WORKFLOW_NODE_TYPES.DELAY:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.delay.default_name"),
          config: defaultNodeConfigForDelay(),
        },
      };

    case WORKFLOW_NODE_TYPES.CONDITION: {
      const branch1 = newNode(WORKFLOW_NODE_TYPES.BRANCHBLOCK, { i18n });
      branch1.data.name = `${branch1.data.name} 1`;
      const branch2 = newNode(WORKFLOW_NODE_TYPES.BRANCHBLOCK, { i18n });
      branch2.data.name = `${branch2.data.name} 2`;

      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.condition.default_name"),
          config: defaultNodeConfigForBranchBlock(),
        },
        blocks: [branch1, branch2],
      };
    }

    case WORKFLOW_NODE_TYPES.BRANCHBLOCK:
      return {
        id: newNodeId(),
        type: type,
        blocks: [],
        data: {
          name: t("workflow_node.branch_block.default_name"),
          config: defaultNodeConfigForBranchBlock(),
        },
      };

    case WORKFLOW_NODE_TYPES.TRYCATCH:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.try_catch.default_name"),
        },
        blocks: [newNode(WORKFLOW_NODE_TYPES.TRYBLOCK, { i18n }), newNode(WORKFLOW_NODE_TYPES.CATCHBLOCK, { i18n })],
      };

    case WORKFLOW_NODE_TYPES.TRYBLOCK:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: "",
        },
        blocks: [],
      };

    case WORKFLOW_NODE_TYPES.CATCHBLOCK:
      return {
        id: newNodeId(),
        type: type,
        blocks: [newNode(WORKFLOW_NODE_TYPES.END, { i18n })],
        data: {
          name: t("workflow_node.catch_block.default_name"),
        },
      };

    case WORKFLOW_NODE_TYPES.BIZ_APPLY:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.apply.default_name"),
          config: defaultNodeConfigForBizApply(),
        },
      };

    case WORKFLOW_NODE_TYPES.BIZ_UPLOAD:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.upload.default_name"),
          config: defaultNodeConfigForBizUpload(),
        },
      };

    case WORKFLOW_NODE_TYPES.BIZ_MONITOR:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.monitor.default_name"),
          config: defaultNodeConfigForBizMonitor(),
        },
      };

    case WORKFLOW_NODE_TYPES.BIZ_DEPLOY:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.deploy.default_name"),
          config: defaultNodeConfigForBizDeploy(),
        },
      };

    case WORKFLOW_NODE_TYPES.BIZ_NOTIFY:
      return {
        id: newNodeId(),
        type: type,
        data: {
          name: t("workflow_node.notify.default_name"),
          config: defaultNodeConfigForBizNotify(),
        },
      };

    default:
      throw new Error("Invalid value of `nodeType`");
  }
};

export const duplicateNode = (node: WorkflowNode, options?: { withCopySuffix?: boolean }) => {
  return duplicateNodes([node], options)[0];
};

export const duplicateNodes = (nodes: WorkflowNode[], options?: { withCopySuffix?: boolean }) => {
  function duplicate(node: WorkflowNode, { withCopySuffix, nodeIdMap }: { withCopySuffix: boolean; nodeIdMap: Map<string, string> }) {
    const { produce } = new Immer({ autoFreeze: false });
    return produce(node, (draft) => {
      draft.data ??= {};
      draft.id = newNodeId();
      draft.data.name = withCopySuffix ? `${draft.data?.name || ""}-copy` : `${draft.data?.name || ""}`;

      nodeIdMap.set(node.id, draft.id); // 原节点 ID 映射到新节点 ID

      if (draft.blocks) {
        draft.blocks = draft.blocks.map((block) => duplicate(block, { withCopySuffix: false, nodeIdMap }));
      }

      if (draft.data?.config) {
        switch (draft.type) {
          case WORKFLOW_NODE_TYPES.BIZ_DEPLOY:
            {
              const prevNodeId = draft.data.config.certificateOutputNodeId as string;
              if (nodeIdMap.has(prevNodeId)) {
                draft.data.config = {
                  ...draft.data.config,
                  certificateOutputNodeId: nodeIdMap.get(prevNodeId),
                };
              }
            }
            break;

          case WORKFLOW_NODE_TYPES.BRANCHBLOCK:
            {
              const stack = [] as Expr[];
              const expr = draft.data.config.expression as Expr;
              if (expr) {
                stack.push(expr);
                while (stack.length > 0) {
                  const n = stack.pop()!;
                  if ("left" in n) {
                    stack.push(n.left);
                    if ("selector" in n.left) {
                      const prevNodeId = n.left.selector.id;
                      if (nodeIdMap.has(prevNodeId)) {
                        n.left.selector.id = nodeIdMap.get(prevNodeId)!;
                      }
                    }
                  }
                  if ("right" in n) {
                    stack.push(n.right);
                  }
                }

                draft.data.config = {
                  ...draft.data.config,
                  expression: expr,
                };
              }
            }
            break;
        }
      }

      return draft;
    });
  }

  const map = new Map<string, string>();
  return nodes.map((node) => duplicate(node, { withCopySuffix: options?.withCopySuffix ?? true, nodeIdMap: map }));
};
// #endregion

// #region Expression
export enum ExprType {
  Constant = "const",
  Variant = "var",
  Comparison = "comparison",
  Logical = "logical",
  Not = "not",
}

export type ExprValue = string | number | boolean;
export type ExprValueType = "string" | "number" | "boolean";
export type ExprValueSelector = {
  id: string;
  name: string;
  type: ExprValueType;
};

export type ExprComparisonOperator = "gt" | "gte" | "lt" | "lte" | "eq" | "neq";
export type ExprLogicalOperator = "and" | "or" | "not";

export type ConstantExpr = { type: ExprType.Constant; value: string; valueType: ExprValueType };
export type VariantExpr = { type: ExprType.Variant; selector: ExprValueSelector };
export type ComparisonExpr = { type: ExprType.Comparison; operator: ExprComparisonOperator; left: Expr; right: Expr };
export type LogicalExpr = { type: ExprType.Logical; operator: ExprLogicalOperator; left: Expr; right: Expr };
export type NotExpr = { type: ExprType.Not; expr: Expr };
export type Expr = ConstantExpr | VariantExpr | ComparisonExpr | LogicalExpr | NotExpr;
// #endregion
